src/compiler/checker.ts

```diff
@@ -14065,6 +14044,126 @@ namespace ts {
+                    }
+                    return anyType;
+                }
+
+                // Resolve the signatures, preferring constructor
+                let signatures = getSignaturesOfType(valueType, SignatureKind.Construct);
+                let ctor = true;
+                if (signatures.length === 0) {
+                    // No construct signatures, try call signatures
+                    signatures = getSignaturesOfType(valueType, SignatureKind.Call);
+                    ctor = false;
+                    if (signatures.length === 0) {
+                        // We found no signatures at all, which is an error
+                        return unknownType;
+                    }
+                }
+
+                return getUnionType(map(signatures, ctor ? getJsxPropsTypeFromConstructSignature : getJsxPropsTypeFromCallSignature), UnionReduction.None);
+            }
+        }
+
+        function getJsxPropsTypeFromCallSignature(sig: Signature) {
+            let propsType = getTypeOfFirstParameterOfSignature(sig);
+            const intrinsicAttribs = getJsxType(JsxNames.IntrinsicAttributes);
+            if (intrinsicAttribs !== unknownType) {
+                propsType = intersectTypes(intrinsicAttribs, propsType);
+            }
+            return propsType;
+        }
+
+        function getJsxPropsTypeFromClassType(hostClassType: Type) {
+            if (isTypeAny(hostClassType)) {
+                return hostClassType;
+            }
+
+            const propsName = getJsxElementPropertiesName();
+            if (propsName === undefined) {
+                // There is no type ElementAttributesProperty, return 'any'
+                return anyType;
+            }
+            else if (propsName === "") {
+                // If there is no e.g. 'props' member in ElementAttributesProperty, use the element class type instead
+                return hostClassType;
+            }
+            else {
+                const attributesType = getTypeOfPropertyOfType(hostClassType, propsName);
+
+                if (!attributesType) {
+                    // There is no property named 'props' on this instance type
+                    return emptyObjectType;
+                }
+                else if (isTypeAny(attributesType)) {
+                    // Props is of type 'any' or unknown
+                    return attributesType;
+                }
+                else {
+                    // Normal case -- add in IntrinsicClassElements<T> and IntrinsicElements
+                    let apparentAttributesType = attributesType;
+                    const intrinsicClassAttribs = getJsxType(JsxNames.IntrinsicClassAttributes);
+                    if (intrinsicClassAttribs !== unknownType) {
+                        const typeParams = getLocalTypeParametersOfClassOrInterfaceOrTypeAlias(intrinsicClassAttribs.symbol);
+                        apparentAttributesType = intersectTypes(
+                            length(typeParams)

```

Comment:
In the previous version if typeParams had length 0, then apparentAttributesType doesn’t change. Now it gets intersected with intrinsicClassAttribs. Is that OK?

Reply:
A length of 0 for type params isn't possible; we don't make a type param list, then. HOWEVER, the real issue that you've called to attention is that if the intrinsic class attributes type has _more than one_ parameter (not correct, but I don't think there's an error for it), this now intersects a class type reference that's missing arguments (instead of nothing), I think it should do our normal type parameter-filling behavior using `fillMissingTypeArguments`. Meaning you should be able to write `interface ElementClass<T, Props = {}> {  /*...*/ }` and actually have it work. I'll add that now. 🌞

Reply:
Done. I think this is now much more consistent with how we handle type arguments and generics elsewhere.